queue/abstract/queue_session.lua

```diff
@@ -0,0 +1,195 @@
+-- If session_uuid == nil - create a new UUID for the connection
+-- and return it, otherwise connect to the existing one.
+local function identify(conn_id, session_uuid)
+    local queue_session_ids = box.space._queue_session_ids
+    local session_ids = queue_session_ids.index.box_sid:get(conn_id)
+    local cur_uuid = session_ids and session_ids[2]
+
+    if session_uuid == nil and cur_uuid == nil then
+        -- Generate new UUID for the session.
+        cur_uuid = uuid.bin()
+        queue_session_ids:insert{conn_id, cur_uuid}
+    elseif session_uuid ~= nil then
+        -- Validate UUID.
+        if not pcall(uuid.frombin, session_uuid) then
+            error('Invalid UUID format.')
+        end
+
+        -- identify using a previously created session.
+        -- Check that a session with this uuid exists.
+        local ids_by_uuid = queue_session_ids.index.uuid:select(
+            session_uuid, { limit = 1 })[1]
+        if ids_by_uuid == nil and inactive_sessions[session_uuid] == nil then
+            error('The UUID ' .. uuid.frombin(session_uuid):str() ..
+                ' is unknown.')
+        end
+
+        if cur_uuid ~= session_uuid then
+            if cur_uuid ~= nil then
+                queue_session.remove_connection(conn_id)
+            end
+            queue_session_ids:insert({conn_id, session_uuid})
+            cur_uuid = session_uuid
+        end
+    end
+
+    -- Exclude the session from inactive.
+    inactive_sessions[cur_uuid] = nil
+
+    return cur_uuid
+end
+
+local function on_session_remove(callback)
+    if type(callback) ~= 'function' then
+        error('The "on_session_remove" argument type must be "function"')
+    end
+    queue_session._on_session_remove = callback
+end
+
+--- Remove a connection from the list of active connections and
+-- release its tasks if necessary.
+local function remove_connection(conn_id)
+    local queue_session_ids = box.space._queue_session_ids
+    local session_uuid = queue_session.get_uuid(conn_id)
+
+    queue_session_ids:delete{conn_id}
+    local session_ids = queue_session_ids.index.uuid:select(session_uuid,
+        {limit = 1})[1]
+
+    -- If a queue session doesn't have any active connections it should be
+    -- removed (if ttr is absent) or moved to the "inactive sessions" list.
+    if session_ids == nil then
+        local ttr = queue_session.settings['ttr'] or 0
+        if ttr > 0 then
+            inactive_sessions[session_uuid] = util.event_time(ttr)
+        elseif queue_session._on_session_remove ~= nil then
+            queue_session._on_session_remove(session_uuid)
+        end
+    end
+end
+
+local function set_settings(opts)
+    opts = opts or {}
+
+    for key, val in pairs(opts) do
+        if key == 'ttr' then
+            if val < 0 then
+                error('Invalid value of ttr: ' .. val)
+            end
+        else
+            error('Unknown option ' .. key)
+        end
+    end
+
+    queue_session.settings = opts
```

Comment:
If I'll do:

```lua
queue_session.set_settings({opt_1 = 'foo"})
queue_session.set_settings({opt_2 = 'bar"})
```

I would expect that the resulting values will be the following:

```lua
{
    opt_1 = 'foo',
    opt_2 = 'bar',
}
```

The same for the corresponding function in abstract.lua.

Reply:
If you want to keep the previous settings: you must get them, change and set them after that.
I think this question will be best to discussed by voice.